types/lodash/index.d.ts

```diff
@@ -6366,98 +6366,34 @@ declare namespace _ {
         /**
          * @see _.forEach
          */
-        each<T>(
-            collection: T[],
-            iteratee?: ListIterator<T, any>
-        ): T[];
-
-        /**
-         * @see _.forEach
-         */
-        each(
-            collection: string,
+        each<TString extends string | null | undefined>(
+            collection: TString,
             iteratee?: ListIterator<string, any>
-        ): string;
+        ): TString;
 
         /**
          * @see _.forEach
          */
-        each<T>(
-            collection: List<T>,
+        each<T, TList extends List<T> | null | undefined>(
```

Comment:
Hmmm... I see what you're saying now. `TList` could be `T[]`, in which case the collection parameter should have type `T[]`, not `List<T>`. I think I agree with you, but I don't think `TList` is quite right either, because collection should never be `null` or `undefined`. Maybe this would be more accurate:

```
each<T, TList extends List<T>, TSource extends TList | null | undefined>(
    collection: TSource,
    iteratee?: ListIterator<T, any, TList>
): TSource;
```

We would need to modify the `ListIterator` type to let the collection type be passed in.